InvenTree/InvenTree/helpers.py

```diff
@@ -542,138 +544,208 @@ def DownloadFile(data, filename, content_type='application/text', inline=False)
     return response
 
 
-def extract_serial_numbers(serials, expected_quantity, next_number: int):
-    """Attempt to extract serial numbers from an input string.
+def increment_serial_number(serial: str):
+    """Given a serial number, (attempt to) generate the *next* serial number.
 
-    Requirements:
-        - Serial numbers can be either strings, or integers
-        - Serial numbers can be split by whitespace / newline / commma chars
-        - Serial numbers can be supplied as an inclusive range using hyphen char e.g. 10-20
-        - Serial numbers can be defined as ~ for getting the next available serial number
-        - Serial numbers can be supplied as <start>+ for getting all expecteded numbers starting from <start>
-        - Serial numbers can be supplied as <start>+<length> for getting <length> numbers starting from <start>
+    Note: This method is exposed to custom plugins.
 
-    Args:
-        serials: input string with patterns
-        expected_quantity: The number of (unique) serial numbers we expect
-        next_number(int): the next possible serial number
+    Arguments:
+        serial: The serial number which should be incremented
+
+    Returns:
+        incremented value, or None if incrementing could not be performed.
     """
-    serials = serials.strip()
 
-    # fill in the next serial number into the serial
-    while '~' in serials:
-        serials = serials.replace('~', str(next_number), 1)
-        next_number += 1
+    from plugin.registry import registry
 
-    # Split input string by whitespace or comma (,) characters
-    groups = re.split(r"[\s,]+", serials)
+    # Ensure we start with a string value
+    if serial is not None:
+        serial = str(serial).strip()
 
-    numbers = []
-    errors = []
+    # First, let any plugins attempt to increment the serial number
+    for plugin in registry.with_mixin('validation'):
+        result = plugin.increment_serial_number(serial)
+        if result is not None:
+            return str(result)
 
-    # Helper function to check for duplicated numbers
-    def add_sn(sn):
-        # Attempt integer conversion first, so numerical strings are never stored
-        try:
-            sn = int(sn)
-        except ValueError:
-            pass
+    # If we get to here, no plugins were able to "increment" the provided serial value
+    # Attempt to perform increment according to some basic rules
+    return increment(serial)
 
-        if sn in numbers:
-            errors.append(_('Duplicate serial: {sn}').format(sn=sn))
-        else:
-            numbers.append(sn)
+
+def extract_serial_numbers(input_string: str, expected_quantity: int, starting_value=None):
```

Comment:
@SchrodingersGat question: could we also accept an int as `input_string`? That would reduce the breakage with the current code.

Reply:
Thanks @SchrodingersGat